ensips/ccip-write/ccip-write.md

```diff
@@ -0,0 +1,342 @@
+
+## Specification
+
+### Database Implementation Considerations
+
+All supported offchain writing calls MUST rely on EIP-5559, specifying the custom error `StorageHandledByOffChainDatabase`. This error MUST return all required values to implement the EIP-712 signature, ensuring domain ownership.
+
+### L2 Implementation Considerations
+
+Following EIP-5559, this specification MUST support implementation on Layer 2 solutions. The specification describes the writing strategy for L2 relying on the custom error `StorageHandledByL2` with the following arguments:
+
+1. **Chain ID**: the chain ID of the L2 network where the contract is deployed.
+2. **Contract Address**: the address of the contract on the L2 network.
+
+### Subdomain registering
+
+**Register parameters**
+
+As the initial step in registering an offchain subdomain, the `registerParams` function has been implemented to support a wide variety of use cases. This function plays a crucial role in creating a flexible and extensible offchain subdomain registration system.
+
+Given that this step relies on the CCIP-Read standard, this interface SHALL only be implemented by the resolver deployed to Ethereum since it won't be gathering the data through the function call on layer 2, but by directly accessing the storage layout of the given contract.
+
+The function has the following signature:
+
+```solidity
+struct RegisterParams {
+    uint256 price;
+    bytes extraData;
+}
+
+function registerParams(
+    bytes memory name,
+    uint256 duration
+)
+    external
+    view
+    returns (RegisterParams memory);
+```
+
+Parameters:
+
+- `name`: The DNS-encoded name to query
+- `duration`: The duration in seconds for the registration
+
+Return:
+
+- `price`: the amount of ETH charger per second
+- `extraData`: any given structure in an ABI encoded format
+
+Sample of callback function used to validate and handle the response returned by the L2 contract:
+
+```solidity
+function registerParamsCallback(
+    bytes[] memory values,
+    bytes memory extraData
+)
+    public
+    pure
+    returns (RegisterParams memory)
+{
+    return abi.decode(values[0], (RegisterParams));
+}
+```
+
+**Register subdomain**
+
+Aiming to integrate with the already existing interface of domain registration, the register function MUST have the following signature:
+
+```solidity
+function register(
+    string calldata name,
+    address owner,
+    uint256 duration,
+    bytes32 secret,
+    address resolver,
+    bytes[] calldata data,
+    bool reverseRecord,
+    uint16 fuses
```

Comment:
```suggestion
    uint16 fuses,
    bytes memory extraData
```

Important for being able to have a more flexible logic or usecases on the registration